### PR TITLE
Prevent api requests from skipping queue

### LINKED ISSUE #2452: API call not stack with the queue from the website.
### Describe the bug

I have a website about 5k DAU ,and api call from another service through api doc , i can see from console these api from another service jump queue and process immediately.

Is this supposed behavior or a bug?how can i fix this ?

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

call api when there is a queue in  website : https://ai-creator.net/aiart ,

I am hosting stable diffusion webui, i am not so sure how  can you debug online.

### Screenshot

_No response_

### Logs

```shell
...
```
image

you can see there are two jobs ,but i have only set concurrency_count=1 in my code.

### System Info

```shell
>>> import gradio as gr
>>> gr.__version__
'3.4b3'
>>>
```

### Severity

serious, but I can work around it

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description
Fixes #2452

Summary:
- Requests made to `/api` endpoint will be blocked if the queue is enabled for that event, unless a special auth token is passed in the header. This is to allow the queue to make requests.
- Developers can keep the api "open" by setting `queue(api_open=True)`. 

Added some unit tests but you can test yourself as well:

```python
import gradio as gr
import time

def iteration(steps):
    for i in range(steps):
       time.sleep(0.5)
       yield i

gr.Interface(iteration,
             inputs=gr.Slider(minimum=1, maximum=10, step=1, value=5),
             outputs=gr.Number()).queue(api_open=False).launch()
```

In a separate process

```python
assert requests.post(
"http://127.0.0.1:7861/api/predict/",
json={"fn_index": 0, "data": [5], "session_hash": 0}).reason == 'Unauthorized'
```

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

File: gradio/blocks.py [hunk 1]
```diff
@@ -1334,3 +1336,8 @@ def startup_events(self):
         if self.enable_queue:
             utils.run_coro_in_background(self._queue.start)
         utils.run_coro_in_background(self.create_limiter)
+
+    def queue_enabled_for_fn(self, fn_index: int):
```
Comment (freddyaboulton) [inline gradio/blocks.py:1340]: Just a refactor to reduce dupe code

File: gradio/routes.py [hunk 2]
```diff
@@ -332,6 +338,15 @@ async def predict(
                         },
                         status_code=500,
                     )
+            if not app.blocks.api_open and app.blocks.queue_enabled_for_fn(
+                body.fn_index
+            ):
+                if f"Bearer {app.queue_token}" != request.headers.get("Authorization"):
```
Comment (freddyaboulton) [inline gradio/routes.py:344]: I think `Bearer {token}` in the `Authorization` header is the standard way of providing these tokens.

File: gradio/blocks.py [hunk 3]
```diff
@@ -463,6 +463,7 @@ def __init__(
         self.share_url = None
         self.width = None
         self.height = None
+        self.api_open = False
```
Comment (abidlabs) [inline gradio/blocks.py:466]: My feeling is that the API should be open by default to preserve the current behavior.

File: gradio/blocks.py [hunk 4]
```diff
@@ -955,12 +957,15 @@ def queue(
             status_update_rate: If "auto", Queue will send status estimations to all clients whenever a job is finished. Otherwise Queue will send status at regular intervals set by this parameter as the number of seconds.
             client_position_to_load_data: Once a client's position in Queue is less that this value, the Queue will collect the input data from the client. You may make this smaller if clients can send large volumes of data, such as video, since the queued data is stored in memory.
             default_enabled: If True, all event listeners will use queueing by default.
+            api_open: If True, the REST routes of the backend will be open, allowing requests made directly to those endpoints to skip the queue.
+            max_size: The maximum number of events the queue will store at any given moment.
```
Comment (abidlabs) [inline gradio/blocks.py:961]: Thanks for adding!

**Comment by abidlabs** [comment]
Thanks @freddyaboulton, this looks really good to me. One addition: can we hide the "view API" page link in the footer if queue is enabled and `api_open` is `False`?

**Comment by abidlabs** [comment]
This is also very timely as we currently have issues with an app on Spaces whose API endpoint is being called from an external endpoint, degrading the experience for regular users of the Space. I'm going to release a beta version so that we can try this out on the Space itself.

**Comment by freddyaboulton** [comment]
@abidlabs I made it so that the api is open by default and api_open will take precedence over show_api if the queue is enabled! Should be good for another look now!

File: gradio/blocks.py [hunk 5]
```diff
@@ -1026,7 +1031,7 @@ def launch(
             ssl_certfile: If a path to a file is provided, will use this as the signed certificate for https. Needs to be provided if ssl_keyfile is provided.
             ssl_keyfile_password: If a password is provided, will use this with the ssl certificate for https.
             quiet: If True, suppresses most print statements.
-            show_api: If True, shows the api docs in the footer of the app. Default True.
+            show_api: If True, shows the api docs in the footer of the app. Default True. If the queue is enabled, then api_open parameter of .queue() will determine if the api docs are shown, independent of the value of api_docs.
```
Comment (abidlabs) [inline gradio/blocks.py:1034]:
```suggestion
            show_api: If True, shows the api docs in the footer of the app. Default True. If the queue is enabled, then api_open parameter of .queue() will determine if the api docs are shown, independent of the value of show_api.
```

**Comment by abidlabs** [comment]
LGTM @freddyaboulton very nice PR!

**Comment by freddyaboulton** [comment]
Thank you for the review @abidlabs !

**Pull request merged**